src/borg/helpers/fs.py

```diff
@@ -161,12 +161,57 @@ def dir_is_tagged(path, exclude_caches, exclude_if_present):
     return tag_names
 
 
-_safe_re = re.compile(r"^((\.\.)?/+)+")
+def make_path_safe(path):
+    """
+    Make path safe by making it relative and normalized.
 
+    `path` is normalized by making it relative, removing
+    consecutive slashes (e.g. '//'), remove '.' elements,
+    and removing trailing slashes.
 
-def make_path_safe(path):
-    """Make path safe by making it relative and local"""
-    return _safe_re.sub("", path) or "."
+    For reasons of security, a ValueError is raised should
+    `path` contain any '..' elements.
+    """
+    path = path.lstrip("/")
+    if path.startswith("../") or "/../" in path or path.endswith("/..") or path == "..":
+        raise ValueError(f"unexpected '..' element in path {path!r}")
+    path = os.path.normpath(path)
+    return path
+
+
+_dotdot_re = re.compile(r"^(\.\./)+")
+
+
+def remove_dotdot_prefixes(path):
+    """
+    Remove '../'s at the beginning of `path`. Additionally,
+    the path is made relative.
+
+    `path` is expected to be normalized already (e.g. via `os.path.normpath()`).
+    """
+    path = path.lstrip("/")
+    path = _dotdot_re.sub("", path)
+    if path in ["", ".."]:
+        return "."
+    return path
+
+
+def assert_normalized_path(path):
+    # `path` should have been normalized earlier. Some features,
+    # like pattern matching rely on a normalized path. As a
+    # precaution we check here again.
+    if to_normalized_path(path) != path:
+        raise ValueError(f"path {path!r} is not normalized")
+    return path
+
+
+def to_normalized_path(path):
+    assert isinstance(path, str)
+    # Legacy versions of Borg still allowed non-normalized paths
+    # to be stored. So, we normalize them when reading.
+    #
+    # Also checks for '..' elements in `path` for reasons of security.
+    return make_path_safe(path)
```

Comment:
btw, as we also use `os.path.normpath` (which implements "normalizing a path"), the names here can be a bit confusing.

Reply:
True, I'll try to come up with a different name. (Suggestions welcome)